client/finality-grandpa/rpc/src/lib.rs

```diff
@@ -19,64 +19,159 @@
+	/// side its justification.
+	#[pubsub(
+		subscription = "grandpa_justifications",
+		subscribe,
+		name = "grandpa_subscribeJustifications"
+	)]
+	fn subscribe_justifications(
+		&self,
+		metadata: Self::Metadata,
+		subscriber: Subscriber<Notification>
+	);
+
+	/// Unsubscribe from receiving notifications about recently finalized blocks.
+	#[pubsub(
+		subscription = "grandpa_justifications",
+		unsubscribe,
+		name = "grandpa_unsubscribeJustifications"
+	)]
+	fn unsubscribe_justifications(
+		&self,
+		metadata: Option<Self::Metadata>,
+		id: SubscriptionId
+	) -> FutureResult<bool>;
 }
 
 /// Implements the GrandpaApi RPC trait for interacting with GRANDPA.
-pub struct GrandpaRpcHandler<AuthoritySet, VoterState> {
+pub struct GrandpaRpcHandler<AuthoritySet, VoterState, Block: BlockT> {
 	authority_set: AuthoritySet,
 	voter_state: VoterState,
+	justification_receiver: GrandpaJustificationReceiver<Block>,
+	manager: SubscriptionManager,
 }
 
-impl<AuthoritySet, VoterState> GrandpaRpcHandler<AuthoritySet, VoterState> {
+impl<AuthoritySet, VoterState, Block: BlockT> GrandpaRpcHandler<AuthoritySet, VoterState, Block> {
 	/// Creates a new GrandpaRpcHander instance.
-	pub fn new(authority_set: AuthoritySet, voter_state: VoterState) -> Self {
+	pub fn new(
+		authority_set: AuthoritySet,
+		voter_state: VoterState,
+		justification_receiver: GrandpaJustificationReceiver<Block>,
+		manager: SubscriptionManager,
+	) -> Self {
 		Self {
 			authority_set,
 			voter_state,
+			justification_receiver,
+			manager,
 		}
 	}
 }
 
-impl<AuthoritySet, VoterState> GrandpaApi for GrandpaRpcHandler<AuthoritySet, VoterState>
+impl<AuthoritySet, VoterState, Block> GrandpaApi<JustificationNotification<Block>>
+	for GrandpaRpcHandler<AuthoritySet, VoterState, Block>
 where
 	VoterState: ReportVoterState + Send + Sync + 'static,
 	AuthoritySet: ReportAuthoritySet + Send + Sync + 'static,
+	Block: BlockT,
 {
+	type Metadata = sc_rpc::Metadata;
+
 	fn round_state(&self) -> FutureResult<ReportedRoundStates> {
 		let round_states = ReportedRoundStates::from(&self.authority_set, &self.voter_state);
 		let future = async move { round_states }.boxed();
 		Box::new(future.map_err(jsonrpc_core::Error::from).compat())
 	}
+
+	fn subscribe_justifications(
+		&self,
+		_metadata: Self::Metadata,
+		subscriber: Subscriber<JustificationNotification<Block>>
+	) {
+		let stream = self.justification_receiver.subscribe()
+			.map(|x| Ok::<_,()>(x))
+			.map_err(|e| warn!("Notification stream error: {:?}", e))
+			.compat();
+
+		self.manager.add(subscriber, |sink| {
+			let stream = stream.map(|res| Ok(res));
+			sink.sink_map_err(|e| warn!("Error sending notifications: {:?}", e))
+				.send_all(stream)
+				.map(|_| ())
+		});
+	}
+
+	fn unsubscribe_justifications(
+		&self,
+		_metadata: Option<Self::Metadata>,
+		id: SubscriptionId
+	) -> FutureResult<bool> {
+		let cancel = self.manager.cancel(id);
+		let future = async move { cancel }
```

Comment:
That seems fairly overcomplicated. Why not `Ok(self.manager.cancel(id))` like we do with every other `unsubscribe` method? Also the specific error seems to be redundant as well.

Reply:
Good point, changing it to return a `Result` instead of the Future`